### PR TITLE
Support specifying upload options for image builds

### PR DESCRIPTION
**Pull request by thozza**
### Notable changes

- Add support for specifying upload options on Koji CLI and thus upload the built image directly to the cloud as part of the build task. One can currently specify only one upload options, which will be used to upload all resulting images (image types x architectures matrix) to the cloud.
- Extend the integration test to test cloud upload with AWS image. The test case fetches relevant information from Koji to identify the image and then checks that it exists in AWS. The image is then deleted.
- Extend unit tests to cover the newly added functionality.

### CODE REVIEW

File: plugins/hub/osbuild.py [hunk 1]
```diff
@@ -95,6 +190,14 @@
                     "type": "object",
                     "$ref": "#/definitions/ostree"
                 },
+                "upload_options": {
+                    "oneOf": [
+                        {"$ref": "#/definitions/AWSEC2UploadOptions"},
+                        {"$ref": "#/definitions/AWSS3UploadOptions"},
+                        {"$ref": "#/definitions/GCPUploadOptions"},
+                        {"$ref": "#/definitions/AzureUploadOptions"}
+                    ]
```
Comment (gicmo) [inline plugins/hub/osbuild.py:199]: I would just route that through and let Composer verify the schema, we do the same for `customizations`. We are not the source of truth here and it will be tedious to keep them in sync.

Comment (thozza) [reply plugins/hub/osbuild.py:199]: ACK

Comment (thozza) [reply plugins/hub/osbuild.py]: removed....

File: test/integration/test_koji.py [hunk 2]
```diff
@@ -37,18 +24,71 @@ def koji_command(*args, _input=None, _globals=None, **kwargs):
...
+        "rhel": [
+            {"url": "http://download.devel.redhat.com/released/RHEL-8/$release/BaseOS/$arch/os/",
+            "package_sets": "blueprint; build; packages"},
+            {"url": "http://download.devel.redhat.com/released/RHEL-8/$release/AppStream/$arch/os/",
+            "package_sets": "blueprint; build; packages"},
+        ]
+    }
+
+    def __init__(self):
+        info = self.parse_os_release()
```
Comment (ondrejbudai) [inline test/integration/test_koji.py:62]: This uses `self.` but `parse_os_release` is a `@staticmethod`. It feels non-idiomatic, but I'm not that experienced in Python to judge.

File: plugins/builder/osbuild.py [hunk 3]
```diff
@@ -184,13 +186,47 @@ def as_dict(self):
...
 
-class ImageStatus(enum.Enum):
-    SUCCESS = "success"
-    FAILURE = "failure"
-    PENDING = "pending"
-    BUILDING = "building"
-    UPLOADING = "uploading"
-    REGISTERING = 'registering'
+class ComposeStatusError:
+    def __init__(self, error_id: int, reason: str, details: Union[Dict, None]):
```
Comment (ondrejbudai) [inline plugins/builder/osbuild.py:240]: Cannot this be `Optional[Dict]`?

File: plugins/builder/osbuild.py [hunk 4]
```diff
@@ -184,13 +186,47 @@ def as_dict(self):
...
+        data = {
+            "id": self.error_id,
+            "reason": self.reason
+        }
+        if self.details:
+            data["details"] = self.details
+        return data
+
+
+class ImageStatus:
```
Comment (ondrejbudai) [inline plugins/builder/osbuild.py:255]: The introduction of this class changes metadata that we save in Brew, right? In that case, it should be a separate commit.

File: plugins/builder/osbuild.py [hunk 5]
```diff
@@ -184,13 +186,47 @@ def as_dict(self):
...
+    def __init__(self, status: str, upload_status: Union[Dict, None], error: Union[ComposeStatusError, None]) -> None:
```
Comment (ondrejbudai) [inline plugins/builder/osbuild.py:256]: `Optional` once again.

File: plugins/builder/osbuild.py [hunk 6]
```diff
@@ -110,13 +110,63 @@ def as_dict(self, arch: str = ""):
...
+        return instance
+
+    def get_options(self, arch: str, image_type: str) -> Union[Dict, None]:
+        if self._common_options is not None:
+            return self._common_options
+
+        if self._arch_img_options_map is None:
+            return None
+
+        try:
```
Comment (ondrejbudai) [inline plugins/builder/osbuild.py:155]:
```python3
arch_options = self._arch_img_options_map.get(arch)
if not arch_options:
    return None
```

doesn't require a `try` block.

File: test/unit/test_builder.py [hunk 7]
```diff
@@ -1258,3 +1312,69 @@ def test_compose_status_retry(self):
...
         client.wait_for_compose(compose_id, sleep_time=0.1)
+
+
+@PluginTest.load_plugin("builder")
+class TestUploadOptionsRegistry(PluginTest):
+
+    def test_no_options(self):
+        """Test that no options are returned if no upload options are set"""
+        reg = self.plugin.UploadOptionsRegistry.from_data(None)
+        for c in itertools.product(["x86_64", "s390x"], ["aws", "azure", "gce", "guest-image"]):
```
Comment (ondrejbudai) [inline test/unit/test_builder.py:1323]: I suggest `for arch, image_type in...`

File: test/unit/test_builder.py [hunk 8]
```diff
@@ -1258,3 +1312,69 @@ def test_compose_status_retry(self):
...
+                "upload_options": {
+                    "region": "us-east-1",
+                    "share_with_accounts": ["123456789"]
+                }
+            },
+            {
+                "arch": "aarch64",
+                "image_type": "aws",
+                "upload_options": {
+                    "region": "us-east-123",
```
Comment (ondrejbudai) [inline test/unit/test_builder.py:1352]: East US is now completely filled with AWS datacenters. 🤭

**Review by ondrejbudai — CHANGES_REQUESTED** [review]
I like this, it feels very clean. Some minor nitpicks inline.

File: plugins/builder/osbuild.py (hunk 3 above)
Comment (thozza) [reply plugins/builder/osbuild.py:240]: yes, it can...

File: test/integration/test_koji.py (hunk 2 above)
Comment (thozza) [reply test/integration/test_koji.py:62]: So static method does not rely on the state of the object nor the class, but it is always an attribute of both. I could reference it using the class name, but it feels silly to me. I'm happy to do a small research to learn what is idiomatic.

File: plugins/builder/osbuild.py (hunk 4 above)
Comment (thozza) [reply plugins/builder/osbuild.py:255]: Can you be more specific about what you mean by metadata? The build task has some associated metadata which are stored with the build (like the content generator that was used, its version, buildroot, etc). The status is uploaded as a JSON file to the build task. Nevertheless I can extract it into a separate commit.

File: plugins/builder/osbuild.py (hunk 5 above)
Comment (thozza) [reply plugins/builder/osbuild.py:256]: _nods_

File: test/unit/test_builder.py (hunk 7 above)
Comment (thozza) [reply test/unit/test_builder.py:1323]: makes sense

File: test/unit/test_builder.py (hunk 8 above)
Comment (thozza) [reply test/unit/test_builder.py:1352]: I know... those datacenters, right? 🤣

File: plugins/builder/osbuild.py (hunk 4 above)
Comment (ondrejbudai) [reply plugins/builder/osbuild.py:255]: Yes, I meant the JSON file that's uploaded. It's now different, isn't it?

File: test/integration/test_koji.py (hunk 2 above)
Comment (thozza) [reply test/integration/test_koji.py:62]: So it seems that calling a static method on the class object is preferred to make it easier to sub-class the class. A class inheriting from this one could override the method and calling it on `self` would result in calling the overridden method. It is not an issue in this case, but I'll rework it.

File: plugins/builder/osbuild.py (hunk 4 above)
Comment (thozza) [reply plugins/builder/osbuild.py:254]: done

Comment (thozza) [reply plugins/builder/osbuild.py:254]: > Yes, I meant the JSON file that's uploaded. It's now different, isn't it?

yes, it is...

**Comment by thozza** [comment]
@ondrejbudai thanks a lot for your review and suggestions. I addressed all of them.

**Review by ondrejbudai — COMMENTED** [review]
Looks good from my point of view. :heart:

**Comment by thozza** [comment]
I removed the arch-specific upload options completely. Based on the discussion that we had so far, it seems that we don't need them at all (at least for now). I kept only the the global upload options.

I defined schema for upload options based on Composer API schema. I excluded the `image_name`, but kept everything that is marked as **required** by the schema. While we could hide more options from the user and have them defined in the Builder configuration, this approach IMO does not make much sense. The reasons from my PoV are:

- I would argue that the Koji integration and API exposed to user can be more low level, since the expectation is that Koji / Brew is used by distribution maintainers / developers. I do not see the value in restricting what they can achieve via API too much.
- Changing values in the Builder configuration would be just an additional step that takes arguably longer than calling the CLI with a different argument.

I do not have strong preference though, since I won't be using the API, so I'll be happy to rework the implementation and move some of the options to Builder configuration. Nevertheless I would welcome some specific requirements or ideas about what would be expected (or how to reach the agreement).

**Review by ondrejbudai — COMMENTED** [review]
Looks good! I would be in favor of adding the image name because I think that pungi should set it to the compose id, so the resulting images are easy to find.

**Comment by thozza** [comment]
> Looks good! I would be in favor of adding the image name because I think that pungi should set it to the compose id, so the resulting images are easy to find.

I added the image name and as a result, the accepted upload options schema is 1:1 copy of composer's cloud API schema.

**Review by ondrejbudai — APPROVED** [review]
